src/Microsoft.DotNet.MSBuildSdkResolver/SemanticVersion.cs

```diff
@@ -0,0 +1,181 @@
+        {
+        }
+
+        public SemanticVersion(int major, int minor, int patch, string pre) :
+            this(major, minor, patch, pre, string.Empty)
+        {
+        }
+
+        public SemanticVersion(int major, int minor, int patch, string pre, string build)
+        {
+            Major = major;
+            Minor = minor;
+            Patch = patch;
+            Pre = pre;
+            Build = build;
+        }
+
+        public static bool operator ==(SemanticVersion s1, SemanticVersion s2)
+        {
+            return Compare(s1, s2) == 0;
+        }
+
+        public static bool operator !=(SemanticVersion s1, SemanticVersion s2)
+        {
+            return !(s1 == s2);
+        }
+
+        public static bool operator <(SemanticVersion s1, SemanticVersion s2)
+        {
+            return Compare(s1, s2) < 0;
+        }
+
+        public static bool operator >(SemanticVersion s1, SemanticVersion s2)
+        {
+            return Compare(s1, s2) > 0;
+        }
+
+        public static bool operator >=(SemanticVersion s1, SemanticVersion s2)
+        {
+            return Compare(s1, s2) >= 0;
+        }
+
+        public static bool operator <=(SemanticVersion s1, SemanticVersion s2)
+        {
+            return Compare(s1, s2) <= 0;
+        }
+
+        public static SemanticVersion Parse(string semanticVersionString)
+        {
+            int majorSeparator = semanticVersionString.IndexOf(".");
+            if (majorSeparator == -1)
+            {
+                return null;
+            }
+
+            int major = 0;
+            if (!int.TryParse(semanticVersionString.Substring(0, majorSeparator), out major))
+            {
+                return null;
+            }
+
+            int minorStart = majorSeparator + 1;
+            int minorSeparator = semanticVersionString.IndexOf(".", minorStart);
+            if (minorSeparator == -1)
+            {
+                return null;
+            }
+
+            int minor = 0;
+            if (!int.TryParse(semanticVersionString.Substring(minorStart, minorSeparator - minorStart), out minor))
+            {
+                return null;
+            }
+
+            int patch = 0;
+            int patchStart = minorSeparator + 1;
+            int patchSeparator = semanticVersionString.FindFirstNotOf("0123456789", patchStart);
+            if (patchSeparator == -1)
+            {
+                if (!int.TryParse(semanticVersionString.Substring(patchStart), out patch))
+                {
+                    return null;
+                }
+
+                return new SemanticVersion(major, minor, patch);
+            }
+
+            if (!int.TryParse(semanticVersionString.Substring(patchStart, patchSeparator - patchStart), out patch))
+            {
+                return null;
+            }
+
+            int preStart = patchSeparator;
+            int preSeparator = semanticVersionString.IndexOf("+", preStart);
```

Comment:
Note that fx_ver used by core-setup doesn't support this. It's fine to do it here, but we'll need to patch core-setup before we can actually ship versions like this.